src/executor/validate/mod.rs

```diff
@@ -0,0 +1,116 @@
+mod unique;
+
+use {
+    crate::{data::Row, result::Result, store::Store},
+    serde::Serialize,
+    sqlparser::ast::{ColumnDef, Ident},
+    std::{fmt::Debug, rc::Rc},
+    thiserror::Error as ThisError,
+    unique::{create_unique_constraints, fetch_all_unique_columns, specified_columns_only},
+};
+
+#[derive(ThisError, Debug, PartialEq, Serialize)]
+pub enum ValidateError {
+    #[error("conflict! storage row has no column on index {0}")]
+    ConflictOnStorageColumnIndex(usize),
+
+    #[error("duplicate entry '{0}' for unique column '{1}'")]
+    DuplicateEntryOnUniqueField(String, String),
+
+    #[error("incompatible type attempted, value: {attempted_value} used for typed column: {column_name} ({column_type})")]
+    IncompatibleTypeOnTypedField {
+        attempted_value: String,
+        column_name: String,
+        column_type: String,
+    },
+}
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+pub enum UniqueKey {
+    Bool(bool),
+    I64(i64),
+    Str(String),
+    Null,
+}
+
+#[derive(Clone)]
+pub enum ColumnValidation {
+    All(Rc<[ColumnDef]>),
+    SpecifiedColumns(Rc<[ColumnDef]>, Vec<Ident>),
+}
+
+pub async fn validate_rows<T: 'static + Debug>(
+    storage: &impl Store<T>,
+    table_name: &str,
+    column_validation: ColumnValidation,
+    row_iter: impl Iterator<Item = &Row> + Clone,
+) -> Result<()> {
+    validate_unique(
+        storage,
+        table_name,
+        column_validation.clone(),
+        row_iter.clone(),
+    )
+    .await?;
+    validate_types(column_validation.clone(), row_iter.clone()).await?;
+    Ok(())
+}
+
+async fn validate_unique<T: 'static + Debug>(
```

Comment:
Let's also move this `validate_unique` into `validate/unique.rs`.
Then, all current codes in `validate/unique.rs` can become private.

Reply:
Done